REGR: fix recursion error in GeoDataFrame constructor with MultiIndex

jorisvandenbossche opened the pull request:
Closes https://github.com/geopandas/geopandas/issues/2486

jorisvandenbossche commented on geopandas/tests/test_geodataframe.py:
```diff
@@ -1220,6 +1220,25 @@ def test_repeat_geo_col(self):
+        # Getting the .geometry column gives GeoDataFrame for both columns
+        # (but with first MultiIndex level removed)
+        # TODO should this give an error instead?
+        result = gdf.geometry
```
The current behaviour is because of

```
    def _get_geometry(self):
        if self._geometry_column_name not in self:
            ... (raise errors)
        return self[self._geometry_column_name]
```

Because `self._geometry_column_name` is "geometry" (and thus found as present), we basically return `gdf["geometry"]`, which in pandas will select out the first MultiIndex level.

This is probably a broader question than this regression fix, since it was already like this before as well (so for a separate issue), but triggered by writing this test, I am wondering:

- Should we maybe (long term) guarantee that `.geometry` always returns a GeoSeries? (so in case it would return a GeoDataFrame, raise an error instead. Similarly as we started raising an error for duplicate "geometry" columns)
- Should the default `_geometry_column_name` actually still be set to "geometry" if that column is not present as a single geometry-dtype column? Maybe we should rather set it to None?

m-richards replied:
Thanks for the PR, I was also having a quick look into this. 
>Should we maybe (long term) guarantee that .geometry always returns a GeoSeries?

I think this makes sense - the natural expectation is that calling .geometry will return something that calling a geometry method like `area`, `x/y` would apply to. We just have to make sure that doing this doesn't break multiindex generated internally by pandas (from e.g. unstack, pivot). Agree that this should be a separate issue / PR (it's generally tied into the less than ideal handling of MultiIndex columns).
>Should the default _geometry_column_name actually still be set to "geometry" if that column is not present as a single geometry-dtype column? Maybe we should rather set it to None?

This sounds reasonable, except perhaps for the case of #2088 where multiindex data is passed to the gdf constructor and the geometry data is added by the `geometry` keyword.

m-richards reviewed (approved):
Looks good, I suppose the only thing to add is the changelog entry, now we are doing that on individual PRs?

Aside, this also appears to fix
#2057 (which I noticed when I was digging into this myself) but `assert_geodataframe_equal` doesn't handle the multiindexes properly and crashes. So perhaps also something for another PR. (This was the test snippet, which I borrowed from #2099)
```python
def test_pivot():
    df = GeoDataFrame(
        {
            "geometry": [Point(x, x) for x in range(3)],
            "value1": np.arange(3, dtype="int64"),
            "value2": np.array([1, 2, 1], dtype="int64"),
        }
    )
    # https://github.com/geopandas/geopandas/issues/2057
    # pivot failing due to creating a MultiIndex
    expected = GeoDataFrame(pd.DataFrame(df.pivot(columns="value1")))
    actual = df.pivot(columns="value1")
    assert_geodataframe_equal(expected, actual) # -> crashes
    assert actual._geometry_column_name == "geometry"
```

jorisvandenbossche commented:
>  I suppose the only thing to add is the changelog entry, now we are doing that on individual PRs?

Right, added!

> Aside, this also appears to fix https://github.com/geopandas/geopandas/issues/2057

Thanks, I added your test for that

jorisvandenbossche commented:
> but `assert_geodataframe_equal` doesn't handle the multiindexes properly and crashes. So perhaps also something for another PR.

Yeah, that's something I generally struggled with here, that `assert_geodataframe_equal` breaks down in several corner cases (MultiIndex, no proper geometry column, ..)

m-richards commented:
Thanks @jorisvandenbossche!
